base_layer/core/src/chain_storage/checkpoint_utils.rs

```diff
@@ -0,0 +1,140 @@
+//  Copyright 2020, The Tari Project
+//
+//  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+//  following conditions are met:
+//
+//  1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+//! ## Common blockchain utility functions for ArrayLike impls
+
+use crate::chain_storage::ChainStorageError;
+use std::cmp;
+use tari_mmr::{ArrayLike, ArrayLikeExt, MerkleCheckPoint};
+
+/// Calculate the total leaf node count upto a specified height. If the height is less than the effective pruning
+/// horizon the total number of nodes up to the effective pruned height is returned.
```

Comment:
It was duplicated in lmdb_db and memory_db. The duplication causes time wasted working out why tests don't work (because you change something in one and not the other) and it's a recipe for bugs. This makes the leak problem smaller by containing it in one place - we need only look at this file to know what operations are used RE checkpoints. 
This module is not public to anything outside of chain storage. Where should I put code that is common to both dbs?

Reply:
I would put this with lmdb, and give memory db access to this. 
The reasoning behind this. Is that memory db is suppose to work like the lmdb, thus sharing it should be okay.

Reply:
I disagree

Reply:
I disagree - but @mikethetike what do you think - I'll do whatever others in the community agree on

Reply:
To be clear, I agree with what you're saying, but I disagree with how to solve it - and with the issue with putting it in the `chain_storage::checkpoint_utils` module (though I guess it depends on if a developer decides to "just use" these functions where ever within `chain_storage` because "it's there").

We should then move `lmdb` and `memory` into `backends`, move the `checkpoint_utils` stuff into `backends` and have it private to `backends`. I'm sure you can see why I chose this way, this PR is already large enough without moving large swaths of code around.

Reply:
I think having it in `checkpoint_utils` is fine. I don't think it's leaking anything that isn't available already. 

> I would put this with lmdb, and give memory db access to this.

I don't agree with this approach. `lmdb` and `memory` should be distinct to the point where we could theoretically have `memory` and `lmdb` feature flags.